lib/fluent/plugin/buffer.rb

```diff
@@ -172,34 +172,46 @@ def metadata(timekey: nil, tag: nil, variables: nil)
       end
 
       # metadata MUST have consistent object_id for each variation
-      # data MUST be Array of serialized events
+      # data MUST be Array of serialized events, or EventStream
       # metadata_and_data MUST be a hash of { metadata => data }
-      def write(metadata_and_data, bulk: false, enqueue: false)
+      def write(metadata_and_data, format: nil, size: nil, enqueue: false)
         return if metadata_and_data.size < 1
         raise BufferOverflowError, "buffer space has too many data" unless storable?
```

Comment:
Incoming data size is out of consideration here.
Buffer plugin will raise BufferOverflowError if its bytesize is larger than configured limit when output plugin will write event streams into buffer chunks.
It is almost same behavior of v0.12 Fluentd buffer plugin.

The only thing to be considered is the case that buffer chunk size limit is enough large in comparison with total size limit. In that case, soft limiting for buffer total size doesn't work well, and buffer may use much larger size than configured total limit.
I think it's better to add a configuration check for total/chunk bytesize configuration to help such case.